src/pycapsule.rs

```diff
@@ -0,0 +1,330 @@
+///     pub val: u32,
+///  }
+///
+///  let r = Python::with_gil(|py| -> PyResult<()> {
+///     let foo = Foo { val: 123 };
+///     let name = CString::new("builtins.capsule").unwrap();
+///
+///     let capsule = PyCapsule::new(py, foo, name.as_ref(), None)?;
+///
+///     let module = PyModule::import(py, "builtins")?;
+///     module.add("capsule", capsule)?;
+///
+///     let cap: &Foo = PyCapsule::import(py, name.as_ref(), false)?;
+///     assert_eq!(cap.val, 123);
+///     Ok(())
+///  });
+///  assert!(r.is_ok());
+/// ```
+#[repr(transparent)]
+pub struct PyCapsule(PyAny);
+
+pyobject_native_type_core!(PyCapsule, ffi::PyCapsule_Type, #checkfunction=ffi::PyCapsule_CheckExact);
+
+struct CapsuleContents<T: 'static, D: FnOnce(T)> {
+    value: T,
+    destructor: D,
+}
+impl<T: 'static, D: FnOnce(T)> CapsuleContents<T, D> {
+    fn new(value: T, destructor: D) -> Self {
+        Self { value, destructor }
+    }
+}
+
+unsafe extern "C" fn capsule_destructor<T: 'static, F: FnOnce(T)>(capsule: *mut ffi::PyObject) {
+    let ptr = ffi::PyCapsule_GetPointer(capsule, ffi::PyCapsule_GetName(capsule));
+    let CapsuleContents { value, destructor } = *Box::from_raw(ptr as *mut CapsuleContents<T, F>);
+    destructor(value)
+}
+
+impl PyCapsule {
+    /// Constructs a new capsule of whose contents are `T` associated with `name`.
+    pub fn new<'py, T: 'static>(py: Python<'py>, value: T, name: &CStr) -> PyResult<&'py Self> {
+        Self::new_with_destructor(py, value, name, std::mem::drop)
+    }
+
+    /// Constructs a new capsule of whose contents are `T` associated with `name`
+    /// Provide a destructor for when `PyCapsule` is destroyed it will be passed the capsule.
+    pub fn new_with_destructor<'py, T: 'static, F: FnOnce(T)>(
+        py: Python<'py>,
+        value: T,
+        name: &CStr,
+        destructor: F,
+    ) -> PyResult<&'py Self> {
+        let val = Box::new(CapsuleContents::new(value, destructor));
+
+        let cap_ptr = unsafe {
+            ffi::PyCapsule_New(
+                Box::into_raw(val) as *mut c_void,
+                name.as_ptr(),
+                Some(capsule_destructor::<T, F>),
+            )
+        };
+        if cap_ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(unsafe { py.from_owned_ptr::<PyCapsule>(cap_ptr) })
+        }
+    }
+
+    /// Import an existing capsule.
+    ///
+    /// The `name` should match the path to module attribute exactly in the form
+    /// of `module.attribute`, which should be the same as the name within the
+    /// capsule. `no_block` indicates to use
+    /// [PyImport_ImportModuleNoBlock()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModuleNoBlock)
+    /// or [PyImport_ImportModule()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModule)
+    /// when accessing the capsule.
+    ///
+    /// ## Safety
+    /// This is unsafe, as there is no guarantee when casting `*mut void` into `T`.
```

Comment:
I think it's more helpful for users to indicate what invariant they must uphold. Also, elsewhere we've done these as only one-level headings:

```suggestion
    /// # Safety
    /// 
    /// It must be known that this capsule's pointer is to an item of type `T`.
```